beam/src/Rhyolite/Task/Beam/Worker.hs

```diff
@@ -0,0 +1,151 @@
+  , be ~ Postgres, f ~ QExpr be (QNested (QNested QBaseScope)))
+  => Connection -- ^ Connection to the database
+  -> DatabaseEntity be db (TableEntity table) -- ^ A table containing embedded tasks.
+  -> f Bool -- ^ A filter for selecting tasks from the table.
+  -> (forall x. Lens' (table x) (Task a b x))
+  -> (a -> Serializable (m (Serializable b))) -- ^ The action, whose output is set as the result of the task.
+  -> Text -- ^ Name of the worker
+  -> m Bool
+taskWorker dbConn table ready field go workerName = do
+  logger <- askLoggerIO
+  checkedOutValue <-
+    -- BEGIN Transaction
+
+    -- Do the following inside a transaction:
+    -- 1. Get the first task that is not currently checked out by any worker, lock this task
+    -- 2. Update this task to reflect that it has been checked out by current worker
+    -- 3. Perform the serializable task
+    withTransactionSerializableRunBeamPostgres dbConn $ do
+      primaryKeyAndInput <- runSelectReturningOne $ select $
+        -- Only lock one row (limit_ 1)
+        lockingAllTablesFor_ PgSelectLockingStrengthUpdate Nothing $ limit_ 1 $ do
+          task <- all_ table
+
+          -- Both task fields should be empty for an unclaimed task
+          -- Also apply any other filters that may have been passed, using ready
+          guard_ $ isNothing_ (task ^. field . taskResult) &&. isNothing_ (task ^. field . taskCheckedOutBy) &&. ready
+
+          -- Return the primary key (task id) along with a custom field that the user asked for.
+          pure (primaryKey task, task ^. field . taskPayload)
+
+      -- In case we did not find any rows, no update SQL will be run
+      -- The row lock that we acquired above will be reset when the transaction ends.
+      forM primaryKeyAndInput $ \(taskId, input) -> do
+        -- Mark the retrieved task as checked out, by the current worker
+        runUpdate $
+          update table
+            (\task -> (task ^. field . taskCheckedOutBy) <-. val_ (Just workerName))
+            (\task -> primaryKey task ==. val_ taskId)
+
+        runSerializableInsideTransaction dbConn (Logger logger) $ (,) taskId <$> go input
+    -- END Transaction, release row lock
+
+  case checkedOutValue of
+    Nothing -> pure False
+    Just (taskId, action) -> do
+      -- Get the followup Serializable
+      followup <- action
+
+      -- BEGIN Transaction
+      withTransactionSerializableRunBeamPostgres dbConn $ do
+        -- Get the result value from the serializable
+        b <- runSerializableInsideTransaction dbConn (Logger logger) followup
+
+        -- Update the task's result field, set checked out field to null
+        runUpdate $ update table
+          (\task -> mconcat
+            [ task ^. field . taskResult <-. val_ (Just b)
+            , task ^. field . taskCheckedOutBy <-. val_ Nothing])
+          (\task -> primaryKey task ==. val_ taskId)
+      -- END Transaction
+
+      pure True
+
+-- | Run a worker thread
```

Comment:
You'll need [blank lines](https://haskell-haddock.readthedocs.io/en/latest/markup.html#markup) between paragraphs or they'll be on the same line when rendered.

Reply:
I have not added this documentation, I just copied it over from the GroundHog version.